### ISSUE TITLE
how to use

### ISSUE THREAD

**Issue by Ertinhow** [Ryanshuai/pubg_derecoil#1]
Please, can you just explain how to use? just a path to begin.... pleaseeee

**Comment by Ryanshuai** [comment]
I have updated the PyTorch models, the code should be working now.
Install Anaconda and other libraries like PyTorch,  pyinput. (Just run 1_start_main.py, if anything missing, install it.)
Try run 1_start_main.py
Ask me if you are in trouble, I am glad to see others using my code.

**Comment by Ryanshuai** [comment]
I have updated the PyTorch models, the code should now work.
Install Anaconda and other libraries such as PyTorch, pyinput. (Just run 1_start_main.py, if anything is missing, install it.)
Try run 1_start_main.py
Ask me if you are in trouble, I am glad to see others using my code.

**Comment by Ryanshuai** [comment]
Change positions in pubg_derecoil/state/position_constant.py  to change the crop position for image detection.
My screen is 21:9, so you have to try to find yours for your screen.
It is exhausting, but it automatically detects your weapon, your scope, so you don't have to use a special mouse for different weapons.
Again, ask me if in trouble.
I will try to modify my code to adapt more

**Comment by Ertinhow** [comment]
stuck on torchvision 
![image](https://user-images.githubusercontent.com/52810249/107134824-51395780-68d4-11eb-8c03-5bb768f6a372.png)

**Comment by Ryanshuai** [comment]
> stuck on torchvision
> ![image](https://user-images.githubusercontent.com/52810249/107134824-51395780-68d4-11eb-8c03-5bb768f6a372.png)

You need to install Cuda, which is a library for deep learning image detection that runs the detection model on the GPU.
I will modify the code so that it can also run on the CPU.
You can decide which to use according to your computer perfermance.
And, have you downloaded deep learning models such as fire_mode.pth.tar to the correct location?

**Comment by Ertinhow** [comment]
I just follow the errors that appeared on runnig 1_start_main.py, and did pip install, I don't know nothing in python...

**Comment by Ertinhow** [comment]
what is the usage to cpu and gpu with this program? is it very high?

**Comment by Ertinhow** [comment]
what the correct location of fire_mode.pth.tar? pls make a tutorial

**Comment by Ryanshuai** [comment]
> what the correct location of fire_mode.pth.tar? pls make a tutorial

ok, when I have time, wait several days.

**Comment by Ertinhow** [comment]
got stuck on this, pls help!!!!!! 
![image](https://user-images.githubusercontent.com/52810249/108615989-06e1cb80-73e8-11eb-9fda-b5613c73d5e7.png)

**Comment by Ryanshuai** [comment]
> t stuck on this, pls help!!!!!!

Seems working, the white area shows the state of the weapon. Since the default firing mode is full, it is displayed as full.
When everything is fine, it will show the scope rate, grip type, muzzle type.

Busy with work these days, I am on adding readme.

**Comment by Ryanshuai** [comment]
README updated, have a try. Good luck.

**Comment by Ertinhow** [comment]
I've readed but it seems so complicated, what I have to change? edit the file position_constant.py? and what make this script start?
it's a lot of information without a basic tutorial for begginers.

**Comment by msdart** [comment]
> README updated, have a try. Good luck.


Change positions in position_constant.py, enter training mode ,change weapons / others. nothing
-full-1---
-full-1---

any idea?

**Comment by Ryanshuai** [comment]
> > README updated, have a try. Good luck.
> 
> Change positions in position_constant.py, enter training mode ,change weapons / others. nothing
> -full-1---
> -full-1---
> 
> any idea?

Were you in the game? The image detector will be monitoring the screen to determine the status.  
The detector will work once when you push the button tab.
Maybe the PUBG window mode need to be the borderless mode

**Comment by Ryanshuai** [comment]
Were you in the game? The image detector will be monitoring the screen to determine the status.
The detector will work once when you push the button tab.
Maybe the PUBG window mode needs to be the borderless mode

**Comment by msdart** [comment]
> Were you in the game? The image detector will be monitoring the screen to determine the status.
> The detector will work once when you push the button tab.
> Maybe the PUBG window mode needs to be the borderless mode

Play in window (noborder) mode.
Temporary images are saved to the folder, but there is no reaction to changing weapons.

**Comment by Ryanshuai** [comment]
> > Were you in the game? The image detector will be monitoring the screen to determine the status.
> > The detector will work once when you push the button tab.
> > Maybe the PUBG window mode needs to be the borderless mode
> 
> Play in window (noborder) mode.
> Temporary images are saved to the folder, but there is no reaction to changing weapons.

I guess nearly done, I did not consider other screen ratios before. 
I will try to add some modules for users to check if the cropping works properly.
When it is not so busy.

**Comment by msdart** [comment]
> > > Were you in the game? The image detector will be monitoring the screen to determine the status.
> > > The detector will work once when you push the button tab.
> > > Maybe the PUBG window mode needs to be the borderless mode
> > 
> > 
> > Play in window (noborder) mode.
> > Temporary images are saved to the folder, but there is no reaction to changing weapons.
> 
> I guess nearly done, I did not consider other screen ratios before.
> I will try to add some modules for users to check if the cropping works properly.
> When it is not so busy.


Ok. I'm waiting.
I would very much like to try out your creation)

**Comment by Ryanshuai** [comment]
> > > > Were you in the game? The image detector will be monitoring the screen to determine the status.
> > > > The detector will work once when you push the button tab.
> > > > Maybe the PUBG window mode needs to be the borderless mode
> > > 
> > > 
> > > Play in window (noborder) mode.
> > > Temporary images are saved to the folder, but there is no reaction to changing weapons.
> > 
> > 
> > I guess nearly done, I did not consider other screen ratios before.
> > I will try to add some modules for users to check if the cropping works properly.
> > When it is not so busy.
> 
> Ok. I'm waiting.
> I would very much like to try out your creation)

readme updated， have a try.

**Comment by msdart** [comment]
> > > > > Were you in the game? The image detector will be monitoring the screen to determine the status.
> > > > > The detector will work once when you push the button tab.
> > > > > Maybe the PUBG window mode needs to be the borderless mode
> > > > 
> > > > 
> > > > Play in window (noborder) mode.
> > > > Temporary images are saved to the folder, but there is no reaction to changing weapons.
> > > 
> > > 
> > > I guess nearly done, I did not consider other screen ratios before.
> > > I will try to add some modules for users to check if the cropping works properly.
> > > When it is not so busy.
> > 
> > 
> > Ok. I'm waiting.
> > I would very much like to try out your creation)
> 
> readme updated， have a try.

Updated coordinates, now shows the type of weapon.
The next question, posture and fire_mode at what coordinates are taken?

**Comment by msdart** [comment]
> > > > > Were you in the game? The image detector will be monitoring the screen to determine the status.
> > > > > The detector will work once when you push the button tab.
> > > > > Maybe the PUBG window mode needs to be the borderless mode
> > > > 
> > > > 
> > > > Play in window (noborder) mode.
> > > > Temporary images are saved to the folder, but there is no reaction to changing weapons.
> > > 
> > > 
> > > I guess nearly done, I did not consider other screen ratios before.
> > > I will try to add some modules for users to check if the cropping works properly.
> > > When it is not so busy.
> > 
> > 
> > Ok. I'm waiting.
> > I would very much like to try out your creation)
> 
> readme updated， have a try.

    *The cropped image will be saved in for_data_check dir*
I have only temp_image, its ok?

**Comment by Ryanshuai** [comment]
> > > > > > Were you in the game? The image detector will be monitoring the screen to determine the status.
> > > > > > The detector will work once when you push the button tab.
> > > > > > Maybe the PUBG window mode needs to be the borderless mode
> > > > > 
> > > > > 
> > > > > Play in window (noborder) mode.
> > > > > Temporary images are saved to the folder, but there is no reaction to changing weapons.
> > > > 
> > > > 
> > > > I guess nearly done, I did not consider other screen ratios before.
> > > > I will try to add some modules for users to check if the cropping works properly.
> > > > When it is not so busy.
> > > 
> > > 
> > > Ok. I'm waiting.
> > > I would very much like to try out your creation)
> > 
> > 
> > readme updated， have a try.
> 
> Updated coordinates, now shows the type of weapon.
> The next question, posture and fire_mode at what coordinates are taken?


Currently, it is not used.
This is a feature I will add in the future.

**Comment by Ryanshuai** [comment]
> > > > > > Were you in the game? The image detector will be monitoring the screen to determine the status.
> > > > > > The detector will work once when you push the button tab.
> > > > > > Maybe the PUBG window mode needs to be the borderless mode
> > > > > 
> > > > > 
> > > > > Play in window (noborder) mode.
> > > > > Temporary images are saved to the folder, but there is no reaction to changing weapons.
> > > > 
> > > > 
> > > > I guess nearly done, I did not consider other screen ratios before.
> > > > I will try to add some modules for users to check if the cropping works properly.
> > > > When it is not so busy.
> > > 
> > > 
> > > Ok. I'm waiting.
> > > I would very much like to try out your creation)
> > 
> > 
> > readme updated， have a try.
> 
> ```
> *The cropped image will be saved in for_data_check dir*
> ```
> 
> I have only temp_image, its ok?

Does it show the scope rate like "x4", muzzle type, grip type?
If everything is displayed, the folder is not needed.

**Comment by Ryanshuai** [comment]
> > > > > > Were you in the game? The image detector will be monitoring the screen to determine the status.
> > > > > > The detector will work once when you push the button tab.
> > > > > > Maybe the PUBG window mode needs to be the borderless mode
> > > > > 
> > > > > 
> > > > > Play in window (noborder) mode.
> > > > > Temporary images are saved to the folder, but there is no reaction to changing weapons.
> > > > 
> > > > 
> > > > I guess nearly done, I did not consider other screen ratios before.
> > > > I will try to add some modules for users to check if the cropping works properly.
> > > > When it is not so busy.
> > > 
> > > 
> > > Ok. I'm waiting.
> > > I would very much like to try out your creation)
> > 
> > 
> > readme updated， have a try.
> 
> ```
> *The cropped image will be saved in for_data_check dir*
> ```
> 
> I have only temp_image, its ok?


And, if everything is correct. Could your share your screen resolution ratio and your  state.position_constants.py?

**Comment by msdart** [comment]
> > > > > > > Were you in the game? The image detector will be monitoring the screen to determine the status.
> > > > > > > The detector will work once when you push the button tab.
> > > > > > > Maybe the PUBG window mode needs to be the borderless mode
> > > > > > 
> > > > > > 
> > > > > > Play in window (noborder) mode.
> > > > > > Temporary images are saved to the folder, but there is no reaction to changing weapons.
> > > > > 
> > > > > 
> > > > > I guess nearly done, I did not consider other screen ratios before.
> > > > > I will try to add some modules for users to check if the cropping works properly.
> > > > > When it is not so busy.
> > > > 
> > > > 
> > > > Ok. I'm waiting.
> > > > I would very much like to try out your creation)
> > > 
> > > 
> > > readme updated， have a try.
> > 
> > 
> > ```
> > *The cropped image will be saved in for_data_check dir*
> > ```
> > 
> > 
> > I have only temp_image, its ok?
> 
> Does it show the scope rate like "x4", muzzle type, grip type?
> If everything is displayed, the folder is not needed.

I can check in more detail later.

**Comment by msdart** [comment]
```
crop_position = {
    'fire_mode': [1330, 1648, 1364, 1676],
    'in_tab': [100, 359, 114, 398],
    'posture': [1005, 705, 1048, 743],
    # 'in_scope': [1669, 1179, 1766, 1208],

    'gun1_name': [96, 1353, 96 + 35, 1353 + 64],
    'gun1_scope': [114, 1585],
    'gun1_muzzle': [247, 1308],
    'gun1_grip': [247, 1415],
    # 'gun1_magazine': [330, 2474],
    'gun1_butt': [247, 1737],

    'gun2_name': [328, 1348, 328 + 35, 1350 + 64],
    'gun2_scope': [345, 1580],
    'gun2_muzzle': [478, 1308],
    'gun2_grip': [478, 1415],
    # 'gun2_magazine': [636, 2474],
    'gun2_butt': [478, 1737]
}
```

1920x1080

**Comment by Ryanshuai** [comment]
> I've readed but it seems so complicated, what I have to change? edit the file position_constant.py? and what make this script start?
> it's a lot of information without a basic tutorial for begginers.

crop_position = {
    'fire_mode': [1330, 1648, 1364, 1676],
    'in_tab': [100, 359, 114, 398],
    'posture': [1005, 705, 1048, 743],
    # 'in_scope': [1669, 1179, 1766, 1208],

    'gun1_name': [96, 1353, 96 + 35, 1353 + 64],
    'gun1_scope': [114, 1585],
    'gun1_muzzle': [247, 1308],
    'gun1_grip': [247, 1415],
    # 'gun1_magazine': [330, 2474],
    'gun1_butt': [247, 1737],

    'gun2_name': [328, 1348, 328 + 35, 1350 + 64],
    'gun2_scope': [345, 1580],
    'gun2_muzzle': [478, 1308],
    'gun2_grip': [478, 1415],
    # 'gun2_magazine': [636, 2474],
    'gun2_butt': [478, 1737]
}
1920x1080

Would you like to try this configuration?

**Comment by msdart** [comment]
> Would you like to try this configuration?

![Screenshot_9](https://user-images.githubusercontent.com/11256316/109175835-423d1b00-7797-11eb-94ec-b974e36ba05f.png)


Please see if everything is displayed correctly or not?
Anti-recoil worked only 2 times ...

**Comment by Ryanshuai** [comment]
> > Would you like to try this configuration?
> 
> ![Screenshot_9](https://user-images.githubusercontent.com/11256316/109175835-423d1b00-7797-11eb-94ec-b974e36ba05f.png)
> 
> Please see if everything is displayed correctly or not?
> Anti-recoil worked only 2 times ...

These two weapons do not support fully automatic shooting mode. Use weapons like AKM, m762,m416, vector.
Only the gun name is correct, I need to check what the detector's input is.
Try to run the python script as an administrator to save the input image to dir for_data_check.

**Comment by msdart** [comment]
> These two weapons do not support fully automatic shooting mode. Use weapons like AKM, m762,m416, vector.
> Only the gun name is correct, I need to check what the detector's input is.
> Try to run the python script as an administrator to save the input image to dir for_data_check.

If you leave the script running and exit the game, the left mouse button still continues to slide down)

Python running as admin, i have only temp_image folder, full size screen
![Screenshot_10](https://user-images.githubusercontent.com/11256316/109178533-0192d100-779a-11eb-97d4-430e25faa6bc.png)

**Comment by Ryanshuai** [comment]
> > These two weapons do not support fully automatic shooting mode. Use weapons like AKM, m762,m416, vector.
> > Only the gun name is correct, I need to check what the detector's input is.
> > Try to run the python script as an administrator to save the input image to dir for_data_check.
> 
> If you leave the script running and exit the game, the left mouse button still continues to slide down)
> 
> Python running as admin, i have only temp_image folder, full size screen
> ![Screenshot_10](https://user-images.githubusercontent.com/11256316/109178533-0192d100-779a-11eb-97d4-430e25faa6bc.png)

Use F12 to temporarily stop sliding down.

Weird... the program works fine on my computer.
![WeChat Screenshot_20210225235711](https://user-images.githubusercontent.com/23020565/109179966-631c6500-77c5-11eb-903f-7d5324881f45.png)

Check if it is in image_detect? Manually create a folder named for_data_check?

**Comment by Ryanshuai** [comment]
> > These two weapons do not support fully automatic shooting mode. Use weapons like AKM, m762,m416, vector.
> > Only the gun name is correct, I need to check what the detector's input is.
> > Try to run the python script as an administrator to save the input image to dir for_data_check.
> 
> If you leave the script running and exit the game, the left mouse button still continues to slide down)
> 
> Python running as admin, i have only temp_image folder, full size screen
> ![Screenshot_10](https://user-images.githubusercontent.com/11256316/109178533-0192d100-779a-11eb-97d4-430e25faa6bc.png)

Are you a Python programmer?
This function in the Detector.py generates the image folder. 
    def im2name(self, im):
        im_cv2 = im
        save_dir = "for_data_check"
        os.makedirs(save_dir, exist_ok=True)
        howMany = len(os.listdir(save_dir))
        if howMany < 10:
            save_path = os.path.join(save_dir, str(howMany) + ".png")
            cv2.imwrite(save_path, im_cv2)
            print(save_path)

Or maybe use an IDE like Pycharm to solve this problem?

**Comment by msdart** [comment]
> Check if it is in image_detect? Manually create a folder named for_data_check?

Create for_data_check - 0 imgs. temp_image - 1 img every TAB

I checked the script from AKA - the recoil decreases if I shoot. If I shoot from scope,  recoil does not decrease.

**Comment by Ryanshuai** [comment]
> > Check if it is in image_detect? Manually create a folder named for_data_check?
> 
> Create for_data_check - 0 imgs. temp_image - 1 img every TAB
> 
> I checked the script from AKA - the recoil decreases if I shoot. If I shoot from scope, recoil does not decrease.

It is because the gun scope is not detected correctly.
Okay, I will try to save the for_data_check image as a temp_image dose. It's too late, another day

**Comment by Ryanshuai** [comment]
> > Check if it is in image_detect? Manually create a folder named for_data_check?
> 
> Create for_data_check - 0 imgs. temp_image - 1 img every TAB
> 
> I checked the script from AKA - the recoil decreases if I shoot. If I shoot from scope, recoil does not decrease.

pip install -r requirements.txt  to install the correct version of cv2 to save the images.

**Comment by Protomove** [comment]
Hello, thank you very much for sharing your code, I am excited to try it, I will try it in a few days

**Comment by Protomove** [comment]
Hello, a question, I can't detect the weapon correctly, what could it be due to?

**Comment by Protomove** [comment]
![1](https://user-images.githubusercontent.com/20619937/132263460-10fb0ff6-07ce-402b-b494-a89aec76ab01.png)
![2](https://user-images.githubusercontent.com/20619937/132263462-bf3c60fa-88f1-4d49-b5a0-48257764d61d.png)

**Comment by Ryanshuai** [comment]
> ion, I can't detect the weapon correctly, what could it be due t

Sorry for the late reply. I am currently busy.
Glad to hear you are interested.

This misdetection is mainly due to the position of the icon, please try to make it accurate according to your screen. (As the README says)

Actually, I am writing a Golang version of derecoil, if you can upgrade any of these two, please pull request.

**Comment by Protomove** [comment]
Thank you very much, I am beginning to learn programming. If I manage to update one, I will inform you. To detect the position of the weapons do you use an application to know the size of the position? my level of english is low sorry hehe
